src/box/applier.cc

```diff
@@ -369,6 +369,189 @@ applier_connection_init(struct iostream *io, const struct uri *uri,
 	}
 }
 
+static void
+applier_run_ballot_triggers(struct applier *applier, bool success)
+{
+	trigger_run(&applier->on_ballot_update, &success);
+}
+
+/**
+ * Perform a IPROTO_VOTE exchange: send a request and decode the response.
+ * Fallback to this method if master isn't avare of "internal.ballot" event.
+ */
+static void
+applier_get_ballot_from_vote(struct iostream *io, struct xrow_header *row,
+			     struct ibuf *ibuf, struct ballot *ballot)
+{
+	RegionGuard guard(&fiber()->gc);
+	xrow_encode_vote(row);
+	coio_write_xrow(io, row);
+	coio_read_xrow(io, ibuf, row);
+	if (row->type == IPROTO_OK) {
+		xrow_decode_ballot_xc(row, ballot);
+		return;
+	}
+	xrow_decode_error(row);
+	struct diag *diag = diag_get();
+	struct error *e = diag_last_error(diag);
+	/*
+	 * Master isn't aware of IPROTO_VOTE request.
+	 * It's OK - we can proceed without it.
+	 */
+	if (box_error_code(e) == ER_UNKNOWN_REQUEST_TYPE) {
+		diag_clear(diag);
+		return;
+	}
+	error_raise(e);
+}
+
+/**
+ * Perform one round of IPROTO_WATCH("internal.ballot") exchange: send the
+ * request / acknowledge that previous notification is accepted, wait for
+ * response.
+ */
+static void
+applier_get_ballot_from_event(struct iostream *io, struct xrow_header *row,
+			      struct ibuf *ibuf, struct ballot *ballot,
+			      int *decoded_fields)
+{
+	RegionGuard guard(&fiber()->gc);
+	xrow_encode_watch_key(row, internal_ballot_key, IPROTO_WATCH);
+	coio_write_xrow(io, row);
+	coio_read_xrow(io, ibuf, row);
+	if (row->type != IPROTO_EVENT) {
```

Comment:
You don't have to handle it, since it's received in a separate connection and iostream.

Anyway, as mentioned above, now I cancel the applier ballot watcher as soon as applier reaches subscribe.

Reply:
I see, so you have 2 connections to the same master. Have you tried to process the ballots in the same main connection? It is not needed after subscribe is done, so I suppose it wouldn't be too complicated? Somewhere in join/fetch_snapshot or around.

Reply:
My first approach tried to do it all in one connection, but I dumped it for the following reasons:
1. Yes, now we only need the ballot during the "initial" stages. Before "follow". But later on we might want to subscribe to ballot for longer periods of time, in which case the event notification might come in the middle of a transaction, breaking transaction boundaries and all that. Besides, the event notification originates from master's tx thread, and data originates from relay thread. I'm not sure using the same connection will work at all.
2. After I unsubscribe from the ballot updates there is no ACK for unsubscribing, so I still have to expect already sent but not yet received notifications. Even after the logic that handles them exits. I'm trying to say you can't wait until you've definitely received all the events. There might be none, or there might be one more coming. So again, other not related to events code has to tolerate their receipt.

I deemed both those reasons ugly enough to move events to a separate connection.

Reply:
Ok, I see, thanks for the explanation.